src/Core/Managed/Shared/Extensibility/Implementation/TaskTimer.cs

```diff
@@ -62,34 +72,50 @@ public bool IsStarted
         /// <param name="elapsed">The task to run.</param>
         public void Start(Func<Task> elapsed)
         {
-            var newTokenSource = new CancellationTokenSource();
+            DelayedWork delayedWork = new DelayedWork(elapsed);
+            DelayedWork previousWork = Interlocked.Exchange(ref this.latestWork, delayedWork);
+
+            if (previousWork != null)
+            {
+                previousWork.IsCancellationRequested = true;
+            }
 
-            TaskEx.Delay(this.Delay, newTokenSource.Token)
-                .ContinueWith(
 #if !NET40
-                async previousTask =>
+            Task delayTask = Task.Delay(this.Delay);
 #else
-                    previousTask =>
+            Task delayTask = TaskEx.Delay(this.Delay);
 #endif
+            delayTask.ContinueWith(
+                    (dlyTask) =>
                     {
-                        CancelAndDispose(Interlocked.CompareExchange(ref this.tokenSource, null, newTokenSource));
+                        Interlocked.CompareExchange(ref this.latestWork, null, delayedWork);
+                        if (delayedWork.IsCancellationRequested)
+                        {
+                            return;
+                        }
+
                         try
                         {
-                            Task task = elapsed();
+                            // Task may (or may not) be executed syncronously
+                            Task task = (delayedWork.RunFunc == null)
+                                                ? null
+                                                : delayedWork.RunFunc();
```

Comment:
The `elapsed` parameter (and the `DelayedWork.ReadFunc` property) should never be `null` here. The original code was [failing fast](http://martinfowler.com/ieeeSoftware/failFast.pdf) by allowing the `NullReferenceException` to be thrown and tracked by the `LogException` call inside the `catch` statement below. I think its important to preserve the original behavior.